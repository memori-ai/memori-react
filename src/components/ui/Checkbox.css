.memori-checkbox {
    display: inline-flex;
    box-sizing: border-box;
    align-items: baseline;
    padding: 0;
    margin: 0;
    color: rgba(0, 0, 0, 0.88);
    cursor: pointer;
    font-family: var(--memori-font-family);
    font-size: 14px;
    line-height: unset;
    list-style: none;
}

.memori-checkbox--input-wrapper {
    position: relative;
    top: 0.2em;
    cursor: pointer;
    white-space: nowrap;
}

.memori-checkbox--inner {
    position: relative;
    top: 0;
    display: block;
    width: 16px;
    height: 16px;
    box-sizing: border-box;
    border: 1px solid #d9d9d9;
    border-radius: 4px;
    border-top-color: rgb(217, 217, 217);
    border-right-color: rgb(217, 217, 217);
    border-bottom-color: rgb(217, 217, 217);
    border-left-color: rgb(217, 217, 217);
    background-color: #fff;
    border-collapse: separate;
    direction: ltr;
    inset-inline-start: 0;
    transition: all 0.3s;
}

.memori-checkbox:not(.memori-checkbox--indeterminate) .memori-checkbox--inner::after {
    position: absolute;
    top: 50%;
    display: table;
    width: 5.7142857142857135px;
    height: 9.142857142857142px;
    box-sizing: border-box;
    border: 2px solid #fff;
    border-top: 0;
    border-inline-start: 0;
    content: '';
    inset-inline-start: 21.5%;
    opacity: 0;
    transform: rotate(45deg) scale(0) translate(-50%, -50%);
    transition: all 0.1s cubic-bezier(0.71, -0.46, 0.88, 0.6), opacity 0.1s;
}

.memori-checkbox--input {
    position: absolute;
    z-index: 1;
    width: 100%;
    height: 100%;
    cursor: pointer;
    inset: 0;
    opacity: 0;
}

.memori-checkbox--text {
    padding-inline-end: 8px;
    padding-inline-start: 8px;
}

.memori-checkbox--disabled,
.memori-checkbox--disabled .memori-checkbox--input-wrapper,
.memori-checkbox--disabled .memori-checkbox--input {
    cursor: not-allowed;
}

.memori-checkbox--disabled .memori-checkbox--inner {
    border-color: #d9d9d9;
    background: rgba(0, 0, 0, 0.04);
}

.memori-checkbox.memori-checkbox--disabled .memori-checkbox--inner::after {
    border-color: rgba(0, 0, 0, 0.25);
}

.memori-checkbox--disabled .memori-checkbox--text {
    color: rgba(0, 0, 0, 0.25);
}

.memori-checkbox--input.memori-checkbox--indeterminate+.memori-checkbox--inner,
.memori-checkbox--input:checked:not(:disabled)+.memori-checkbox--inner {
    z-index: 2;
    border-color: var(--memori-primary);
    background-color: var(--memori-primary);
}

.memori-checkbox--input:checked+.memori-checkbox--inner::after {
    opacity: 1;
    transform: rotate(45deg) scale(1) translate(-50%, -50%);
    transition: all 0.2s cubic-bezier(0.12, 0.4, 0.29, 1.46) 0.1s;
}

.memori-checkbox.memori-checkbox--indeterminate .memori-checkbox--inner::after,
.memori-checkbox.memori-checkbox--indeterminate .memori-checkbox--input:checked+.memori-checkbox--inner::after {
    position: absolute;
    top: calc(50% - 1px);
    left: 2px;
    width: 10px;
    border-bottom: 2px solid #fff;
    content: "";
    transform: none;
    transition: all 0.2s cubic-bezier(0.12, 0.4, 0.29, 1.46) 0.1s;
}