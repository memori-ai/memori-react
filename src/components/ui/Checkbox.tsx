import React, { FC, useEffect, useRef } from 'react';
import cx from 'classnames';

export interface Props extends React.HTMLAttributes<HTMLInputElement> {
  label?: string;
  checked?: boolean;
  disabled?: boolean;
  indeterminate?: boolean;
  name?: string;
  onChange?: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

const Checkbox: FC<Props> = ({
  label,
  className,
  disabled = false,
  indeterminate = false,
  checked,
  children,
  onChange,
  name,
  ...props
}) => {
  const inputRef = useRef<HTMLInputElement>(null);

  useEffect(() => {
    if (!inputRef.current) return;
    inputRef.current.indeterminate = indeterminate;
  }, [inputRef, indeterminate]);

  return (
    <label
      className={cx('memori-checkbox', className, {
        'memori-checkbox--disabled': disabled,
        'memori-checkbox--indeterminate': indeterminate,
      })}
    >
      <span className="memori-checkbox--input-wrapper">
        <input
          ref={inputRef}
          type="checkbox"
          {...props}
          name={name}
          disabled={disabled}
          checked={checked || indeterminate}
          className="memori-checkbox--input"
          onChange={onChange}
        />
        <span className="memori-checkbox--inner" />
      </span>
      {!!label?.length && (
        <span className="memori-checkbox--text">{label}</span>
      )}
    </label>
  );
};

export default Checkbox;
