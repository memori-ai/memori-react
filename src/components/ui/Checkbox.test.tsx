import React from 'react';
import { render } from '@testing-library/react';
import Checkbox from './Checkbox';

it('renders Checkbox unchanged', () => {
  const { container } = render(<Checkbox label="Check me" />);
  expect(container).toMatchSnapshot();
});

it('renders Checkbox checked unchanged', () => {
  const { container } = render(<Checkbox label="Check me" checked />);
  expect(container).toMatchSnapshot();
});

it('renders Checkbox disabled unchanged', () => {
  const { container } = render(<Checkbox label="Check me" disabled />);
  expect(container).toMatchSnapshot();
});

it('renders Checkbox checked and disabled unchanged', () => {
  const { container } = render(<Checkbox label="Check me" checked disabled />);
  expect(container).toMatchSnapshot();
});

it('renders Checkbox indeterminate unchanged', () => {
  const { container } = render(<Checkbox label="Check me" indeterminate />);
  expect(container).toMatchSnapshot();
});
