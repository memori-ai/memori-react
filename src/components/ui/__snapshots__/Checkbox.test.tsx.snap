// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Checkbox checked and disabled unchanged 1`] = `
<div>
  <label
    class="memori-checkbox memori-checkbox--disabled"
  >
    <span
      class="memori-checkbox--input-wrapper"
    >
      <input
        checked=""
        class="memori-checkbox--input"
        disabled=""
        type="checkbox"
      />
      <span
        class="memori-checkbox--inner"
      />
    </span>
    <span
      class="memori-checkbox--text"
    >
      Check me
    </span>
  </label>
</div>
`;

exports[`renders Checkbox checked unchanged 1`] = `
<div>
  <label
    class="memori-checkbox"
  >
    <span
      class="memori-checkbox--input-wrapper"
    >
      <input
        checked=""
        class="memori-checkbox--input"
        type="checkbox"
      />
      <span
        class="memori-checkbox--inner"
      />
    </span>
    <span
      class="memori-checkbox--text"
    >
      Check me
    </span>
  </label>
</div>
`;

exports[`renders Checkbox disabled unchanged 1`] = `
<div>
  <label
    class="memori-checkbox memori-checkbox--disabled"
  >
    <span
      class="memori-checkbox--input-wrapper"
    >
      <input
        class="memori-checkbox--input"
        disabled=""
        type="checkbox"
      />
      <span
        class="memori-checkbox--inner"
      />
    </span>
    <span
      class="memori-checkbox--text"
    >
      Check me
    </span>
  </label>
</div>
`;

exports[`renders Checkbox indeterminate unchanged 1`] = `
<div>
  <label
    class="memori-checkbox memori-checkbox--indeterminate"
  >
    <span
      class="memori-checkbox--input-wrapper"
    >
      <input
        checked=""
        class="memori-checkbox--input"
        type="checkbox"
      />
      <span
        class="memori-checkbox--inner"
      />
    </span>
    <span
      class="memori-checkbox--text"
    >
      Check me
    </span>
  </label>
</div>
`;

exports[`renders Checkbox unchanged 1`] = `
<div>
  <label
    class="memori-checkbox"
  >
    <span
      class="memori-checkbox--input-wrapper"
    >
      <input
        class="memori-checkbox--input"
        type="checkbox"
      />
      <span
        class="memori-checkbox--inner"
      />
    </span>
    <span
      class="memori-checkbox--text"
    >
      Check me
    </span>
  </label>
</div>
`;
